gax/src/main/java/com/google/api/gax/rpc/mtls/MtlsProvider.java

```diff
@@ -0,0 +1,195 @@
+  private ProcessProvider processProvider;
+
+  /**
+   * The policy for mutual TLS endpoint usage. NEVER means always use regular endpoint; ALWAYS means
+   * always use mTLS endpoint; AUTO means auto switch to mTLS endpoint if client certificate exists
+   * and should be used.
+   */
+  public enum MtlsEndpointUsagePolicy {
+    NEVER,
+    AUTO,
+    ALWAYS;
+  }
+
+  @VisibleForTesting
+  MtlsProvider(
+      EnvironmentProvider envProvider, ProcessProvider processProvider, String metadataPath) {
+    this.envProvider = envProvider;
+    this.processProvider = processProvider;
+    this.metadataPath = metadataPath;
+  }
+
+  public MtlsProvider() {
+    this(
+        new SystemEnvironmentProvider(),
+        new DefaultProcessProvider(),
+        DEFAULT_CONTEXT_AWARE_METADATA_PATH);
+  }
+
+  /**
+   * Returns if mutual TLS client certificate should be used. If the value is true, the key store
+   * from {@link #getKeyStore()} will be used to configure mutual TLS transport.
+   */
+  public boolean useMtlsClientCertificate() {
+    String useClientCertificate = envProvider.getenv("GOOGLE_API_USE_CLIENT_CERTIFICATE");
+    return "true".equals(useClientCertificate);
+  }
+
+  /** Returns the current mutual TLS endpoint usage policy. */
+  public MtlsEndpointUsagePolicy getMtlsEndpointUsagePolicy() {
+    String mtlsEndpointUsagePolicy = envProvider.getenv("GOOGLE_API_USE_MTLS_ENDPOINT");
+    if ("never".equals(mtlsEndpointUsagePolicy)) {
+      return MtlsEndpointUsagePolicy.NEVER;
+    } else if ("always".equals(mtlsEndpointUsagePolicy)) {
+      return MtlsEndpointUsagePolicy.ALWAYS;
+    }
+    return MtlsEndpointUsagePolicy.AUTO;
+  }
+
+  /** The mutual TLS key store created with the default client certificate on device. */
+  public KeyStore getKeyStore() throws IOException {
+    try (InputStream stream = new FileInputStream(metadataPath)) {
+      return getKeyStore(stream, processProvider);
+    } catch (FileNotFoundException exception) {
+      // If the metadata file doesn't exist, then there is no key store, just return null.
+      return null;
+    }
+  }
+
+  @VisibleForTesting
+  static KeyStore getKeyStore(InputStream metadata, ProcessProvider processProvider)
+      throws IOException {
+    try {
+      Process process = processProvider.createProcess(metadata);
+
+      // Run the command and timeout after 1000 milliseconds.
+      int exitCode = runCertificateProviderCommand(process, 1000);
+      if (exitCode != 0) {
+        throw new IOException("Cert provider command failed with exit code: " + exitCode);
+      }
+
+      // Create mTLS key store with the input certificates from shell command.
+      return SecurityUtils.createMtlsKeyStore(process.getInputStream());
+    } catch (InterruptedException e) {
+      throw new IOException("Interrupted executing certificate provider command", e);
+    } catch (GeneralSecurityException e) {
+      throw new IOException("Failed to get key store", e);
```

Comment:
as above, consider not wrapping this here

Reply:
Done.